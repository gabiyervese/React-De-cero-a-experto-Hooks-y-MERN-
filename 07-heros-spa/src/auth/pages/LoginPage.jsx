import { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { AuthContext } from "../context/AuthContext";

export const LoginPage = () => {
  const navigate = useNavigate();

  const { login } = useContext(AuthContext);

  const onLogin = () => {
    //Aqui obtengo la ultima ruta visitada del local Storage y si no hubo nada coloco por default "/".
    const lastPath = localStorage.getItem("lastPath") || "/";

    login("Gabriela Yervese");

    //Aqui al hacer login en vez de redirigirlo directamente al "/", utilizo la ultima ruta como direccion.
    navigate(lastPath, {
      //usamos el "replace" en true para reemplazar el historial porque no queremos que el user pueda retroceder al login si ya se ha logueado.
      replace: true,
    });
  };

  return (
    <div className="container mt-5">
      <h1>Login</h1>
      <hr />

      <button className="btn btn-primary" onClick={onLogin}>
        Login
      </button>
    </div>
  );
};
