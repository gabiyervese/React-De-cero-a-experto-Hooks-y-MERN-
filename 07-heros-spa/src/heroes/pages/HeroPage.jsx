import { useParams, Navigate, useNavigate } from "react-router-dom";
import { getHeroById } from "../helpers";
import { useMemo } from "react";
import "animate.css";

export const HeroPage = () => {
  const { heroId } = useParams();
  const navigate = useNavigate();

  //Aqui memorizamos la info sobre los heroes que pedimos a la funcion, de esta forma no lo cargamos cada vez. A menos que el heroId cambie.
  const hero = useMemo(() => getHeroById(heroId), [heroId]);

  const onReturn = () => {
    //usar el -1 te lleva a la pantalla del historial anterior.
    navigate(-1);
  };

  //Uso esta redireccion para que si el user manipula la url a una no valida/ que no exista no me aparezca undefined sino que se redirija a esta ruta que yo defini.
  if (!hero) {
    return <Navigate to="/marvel" />;
  }

  return (
    <div className="row mt-5 animate__animated animate__fadeInLeft">
      <div className="col-4">
        <img
          src={`/heroes/${heroId}.jpg`}
          alt={hero.superhero}
          className="img-thumbnail "
        />
      </div>

      <div className="col-8">
        <h3>{hero.superhero}</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item">
            <b>Alter ego:</b>
            {hero.alter_ego}
          </li>
          <li className="list-group-item">
            <b>Publisher:</b>
            {hero.publisher}
          </li>
          <li className="list-group-item">
            <b>First appearance:</b>
            {hero.first_appearance}
          </li>
        </ul>

        <h5 className="mt-3">Characters</h5>
        <p>{hero.characters}</p>

        <button className="btn btn-outline-primary" onClick={onReturn}>
          Return
        </button>
      </div>
    </div>
  );
};
