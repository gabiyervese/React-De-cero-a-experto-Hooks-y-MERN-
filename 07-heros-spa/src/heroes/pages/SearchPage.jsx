import { useLocation, useNavigate } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import { HeroCard } from "../components";
import queryString from "query-string";
import { getHeroByName } from "../helpers/getHeroByName";

export const SearchPage = () => {
  const navigate = useNavigate();
  const location = useLocation();

  //Puedo hacerlo asi o haciendo destructuring del "query"
  //Forma #1
  //const query = queryString.parse(location.search);

  //Forma #2 - recomendada / sacamos el valor de "q" con el destructuring y si no viene nada en la "q" colcoamos un string vacio, es decir, nada.
  const { q = "" } = queryString.parse(location.search);

  const heroes = getHeroByName(q);

  const { searchText, onInputChange } = useForm({
    //colocp "q" en vez de un string vacio para que cuando haga refresh en la pagina no se pierda lo escrito en el input por el user, por lo que se toma la ultima info de q como referencia.
    searchText: q,
  });

  const showSearch = q.length === 0;
  const showError = q.length > 0 && heroes.length === 0;

  const onSearchSubmit = (event) => {
    event.preventDefault();
    //if (searchText.trim().length <= 1) return;

    //Asi enviamos la informacion que el user escribe en el input la cual estara en el "searchText" en los querys params. si colocamos navigate(``) estariamos apuntando al lugar en el que nos encontramos en ese momento.
    //Pero como queremos enviar la info mandamos la query asi... Tambien podemos controlar su forma con lower/upper camel case.
    navigate(`?q=${searchText}`);
  };

  return (
    <>
      <h1>Search</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Searching</h4>
          <hr />
          <form onSubmit={onSearchSubmit} aria-label="form">
            <input
              type="text"
              placeholder="Search a hero"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={onInputChange}
            />
            <button className="btn btn-outline-primary mt-3">Search</button>
          </form>
        </div>

        <div className="col-7">
          <h4>Results</h4>
          <hr />

          {/* NOTA: Esta seria una forma de mostrar o no las alertas */}
          {/* {q === "" ? (
            <div className="alert alert-primary">Search a hero</div>
          ) : (
            heroes.length === 0 && (
              <div className="alert alert-danger">There's no {q} results </div>
            )
          )} */}

          {/* Esta forma de  mostrarlas es mas legible */}
          <div
            className="alert alert-primary"
            style={{ display: showSearch ? "" : "none" }}
            aria-label="search a hero"
          >
            Search a hero
          </div>

          <div
            className="alert alert-danger"
            style={{ display: showError ? "" : "none" }}
            aria-label="no hero found"
          >
            There's no {q} results{" "}
          </div>

          {heroes.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </>
  );
};
