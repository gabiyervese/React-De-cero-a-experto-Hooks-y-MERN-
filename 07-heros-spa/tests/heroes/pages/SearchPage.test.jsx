import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";

const mockedUseNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),

  useNavigate: () => mockedUseNavigate,
}));

beforeEach(() => jest.clearAllMocks());

describe("Pruebas en <SearchPage/>", () => {
  test("Debe de mostrarse correctamente con valores por default", () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    );

    //screen.debug();

    expect(container).toMatchSnapshot();
  });

  test("Debe de mostrar a Batman y el input con el valor del queryString", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <SearchPage />
      </MemoryRouter>
    );

    //screen.debug();

    const input = screen.getByRole("textbox");

    //Confirmamos que el value del input sea batman como en la query string
    expect(input.value).toBe("batman");

    const img = screen.getByRole("img");

    expect(img.src).toContain("/heroes/dc-batman.jpg");
  });

  test("Debe de mostrar el style del div 'search a hero' como none cuando haya un valor en la query string", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <SearchPage />
      </MemoryRouter>
    );

    const searchHeroDiv = screen.getByLabelText("search a hero");

    //screen.debug();

    //Confirmo que este div tenga la clase display none mientras haya un valor en la query string.
    //PD: searchHeroDiv.style._values.display === searchHeroDiv.style.display
    expect(searchHeroDiv.style.display).toBe("none");
  });

  test("Debe de mostrar un error si no se encuentra el heroe", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=fegwef"]}>
        <SearchPage />
      </MemoryRouter>
    );

    const noHeroFoundDiv = screen.getByLabelText("no hero found");

    expect(noHeroFoundDiv.style.display).toBe("");

    //screen.debug();
  });

  test("Debe de llamar el navigate hacia la pantalla nueva", () => {
    render(
      <MemoryRouter initialEntries={["/search"]}>
        <SearchPage />
      </MemoryRouter>
    );

    const input = screen.getByRole("textbox");
    //Simulamos que escribimos superman en el input
    fireEvent.change(input, {
      target: { name: "searchText", value: "superman" },
    });

    const form = screen.getByLabelText("form");
    //Simulamos que enviamos el form
    fireEvent.submit(form);

    //Confirmamos que el useNavigate sea llamado con su respectivo argumento el cual seria la query string "superman"
    expect(mockedUseNavigate).toHaveBeenCalledWith("?q=superman");
  });
});
