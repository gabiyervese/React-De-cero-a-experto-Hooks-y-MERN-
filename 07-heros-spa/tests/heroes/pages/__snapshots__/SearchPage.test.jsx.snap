// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage/> Debe de mostrarse correctamente con valores por default 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-3"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="search a hero"
        class="alert alert-primary"
      >
        Search a hero
      </div>
      <div
        aria-label="no hero found"
        class="alert alert-danger"
        style="display: none;"
      >
        There's no 
         results
         
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Batman"
                class="card-img"
                src="/heroes/dc-batman.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Batman
                </h5>
                <p
                  class="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Superman"
                class="card-img"
                src="/heroes/dc-superman.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Superman
                </h5>
                <p
                  class="card-text"
                >
                  Kal-El
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    Action Comics #1
                  </small>
                </p>
                <a
                  href="/hero/dc-superman"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Flash"
                class="card-img"
                src="/heroes/dc-flash.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Flash
                </h5>
                <p
                  class="card-text"
                >
                  Jay Garrick
                </p>
                <p
                  class="card-text"
                >
                  Jay Garrick, Barry Allen, Wally West, Bart Allen
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    Flash Comics #1
                  </small>
                </p>
                <a
                  href="/hero/dc-flash"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Green Lantern"
                class="card-img"
                src="/heroes/dc-green.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Green Lantern
                </h5>
                <p
                  class="card-text"
                >
                  Alan Scott
                </p>
                <p
                  class="card-text"
                >
                  Alan Scott, Hal Jordan, Guy Gardner, John Stewart, Kyle Raynor, Jade, Sinestro, Simon Baz
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    All-American Comics #16
                  </small>
                </p>
                <a
                  href="/hero/dc-green"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Green Arrow"
                class="card-img"
                src="/heroes/dc-arrow.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Green Arrow
                </h5>
                <p
                  class="card-text"
                >
                  Oliver Queen
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    More Fun Comics #73
                  </small>
                </p>
                <a
                  href="/hero/dc-arrow"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Wonder Woman"
                class="card-img"
                src="/heroes/dc-wonder.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Wonder Woman
                </h5>
                <p
                  class="card-text"
                >
                  Princess Diana
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    All Star Comics #8
                  </small>
                </p>
                <a
                  href="/hero/dc-wonder"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Martian Manhunter"
                class="card-img"
                src="/heroes/dc-martian.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Martian Manhunter
                </h5>
                <p
                  class="card-text"
                >
                  J'onn J'onzz
                </p>
                <p
                  class="card-text"
                >
                  Martian Manhunter
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    Detective Comics #225
                  </small>
                </p>
                <a
                  href="/hero/dc-martian"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Robin/Nightwing"
                class="card-img"
                src="/heroes/dc-robin.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Robin/Nightwing
                </h5>
                <p
                  class="card-text"
                >
                  Dick Grayson
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    Detective Comics #38
                  </small>
                </p>
                <a
                  href="/hero/dc-robin"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Blue Beetle"
                class="card-img"
                src="/heroes/dc-blue.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Blue Beetle
                </h5>
                <p
                  class="card-text"
                >
                  Dan Garret
                </p>
                <p
                  class="card-text"
                >
                  Dan Garret, Ted Kord, Jaime Reyes
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    Mystery Men Comics #1
                  </small>
                </p>
                <a
                  href="/hero/dc-blue"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Black Canary"
                class="card-img"
                src="/heroes/dc-black.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Black Canary
                </h5>
                <p
                  class="card-text"
                >
                  Dinah Drake
                </p>
                <p
                  class="card-text"
                >
                  Dinah Drake, Dinah Lance
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    Flash Comics #86
                  </small>
                </p>
                <a
                  href="/hero/dc-black"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Spider Man"
                class="card-img"
                src="/heroes/marvel-spider.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Spider Man
                </h5>
                <p
                  class="card-text"
                >
                  Peter Parker
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    Amazing Fantasy #15
                  </small>
                </p>
                <a
                  href="/hero/marvel-spider"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Captain America"
                class="card-img"
                src="/heroes/marvel-captain.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Captain America
                </h5>
                <p
                  class="card-text"
                >
                  Steve Rogers
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    Captain America Comics #1
                  </small>
                </p>
                <a
                  href="/hero/marvel-captain"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Iron Man"
                class="card-img"
                src="/heroes/marvel-iron.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Iron Man
                </h5>
                <p
                  class="card-text"
                >
                  Tony Stark
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    Tales of Suspense #39
                  </small>
                </p>
                <a
                  href="/hero/marvel-iron"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Thor"
                class="card-img"
                src="/heroes/marvel-thor.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Thor
                </h5>
                <p
                  class="card-text"
                >
                  Thor Odinson
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    Journey into Myster #83
                  </small>
                </p>
                <a
                  href="/hero/marvel-thor"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Hulk"
                class="card-img"
                src="/heroes/marvel-hulk.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Hulk
                </h5>
                <p
                  class="card-text"
                >
                  Bruce Banner
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    The Incredible Hulk #1
                  </small>
                </p>
                <a
                  href="/hero/marvel-hulk"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Wolverine"
                class="card-img"
                src="/heroes/marvel-wolverine.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Wolverine
                </h5>
                <p
                  class="card-text"
                >
                  James Howlett
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    The Incredible Hulk #180
                  </small>
                </p>
                <a
                  href="/hero/marvel-wolverine"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Daredevil"
                class="card-img"
                src="/heroes/marvel-daredevil.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Daredevil
                </h5>
                <p
                  class="card-text"
                >
                  Matthew Michael Murdock
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    Daredevil #1
                  </small>
                </p>
                <a
                  href="/hero/marvel-daredevil"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Hawkeye"
                class="card-img"
                src="/heroes/marvel-hawkeye.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Hawkeye
                </h5>
                <p
                  class="card-text"
                >
                  Clinton Francis Barton
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    Tales of Suspense #57
                  </small>
                </p>
                <a
                  href="/hero/marvel-hawkeye"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Cyclops"
                class="card-img"
                src="/heroes/marvel-cyclops.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Cyclops
                </h5>
                <p
                  class="card-text"
                >
                  Scott Summers
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    X-Men #1
                  </small>
                </p>
                <a
                  href="/hero/marvel-cyclops"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Silver Surfer"
                class="card-img"
                src="/heroes/marvel-silver.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Silver Surfer
                </h5>
                <p
                  class="card-text"
                >
                  Norrin Radd
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    The Fantastic Four #48
                  </small>
                </p>
                <a
                  href="/hero/marvel-silver"
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
